spec/src/main/resources/META-INF/smithy/bsp/bsp.smithy

```diff
@@ -0,0 +1,1380 @@
+    /// The target’s Uri
+    uri: URI
+}
+
+list BuildTargetIdentifiers {
+    member: BuildTargetIdentifier
+}
+
+timestamp EventTime
+integer DurationMillis
+
+/// Build target contains metadata about an artifact (for example library, test, or binary artifact). Using vocabulary of other build tools:
+///
+/// * sbt: a build target is a combined project + config. Example:
+/// * a regular JVM project with main and test configurations will have 2 build targets, one for main and one for test.
+/// * a single configuration in a single project that contains both Java and Scala sources maps to one BuildTarget.
+/// * a project with crossScalaVersions 2.11 and 2.12 containing main and test configuration in each will have 4 build targets.
+/// * a Scala 2.11 and 2.12 cross-built project for Scala.js and the JVM with main and test configurations will have 8 build targets.
+/// * Pants: a pants target corresponds one-to-one with a BuildTarget
+/// * Bazel: a bazel target corresponds one-to-one with a BuildTarget
+///
+/// The general idea is that the BuildTarget data structure should contain only information that is fast or cheap to compute.
+structure BuildTarget {
+    /// The target’s unique identifier
+    @required
+    id: BuildTargetIdentifier
+    /// A human readable name for this target.
+    /// May be presented in the user interface.
+    /// Should be unique if possible.
+    /// The id.uri is used if None.
+    displayName: String
+    /// The directory where this target belongs to. Multiple build targets are allowed to map
+    /// to the same base directory, and a build target is not required to have a base directory.
+    /// A base directory does not determine the sources of a target, see buildTarget/sources. */
+    baseDirectory: URI
+    /// Free-form string tags to categorize or label this build target.
+    /// For example, can be used by the client to:
+    /// - customize how the target should be translated into the client's project model.
+    /// - group together different but related targets in the user interface.
+    /// - display icons or colors in the user interface.
+    /// Pre-defined tags are listed in `BuildTargetTag` but clients and servers
+    /// are free to define new tags for custom purposes.
+    @required
+    tags: BuildTargetTags
+    /// The set of languages that this target contains.
+    /// The ID string for each language is defined in the LSP.
+    @required
+    languageIds: LanguageIds
+    /// The direct upstream build target dependencies of this build target
+    @required
+    dependencies: BuildTargetIdentifiers
+    /// The capabilities of this build target.
+    @required
+    capabilities: BuildTargetCapabilities
+    /// Kind of data to expect in the `data` field. If this field is not set, the kind of data is not specified.
+    @required
+    dataKind: BuildTargetDataKind
+    /// Language-specific metadata about this target.
+    /// See ScalaBuildTarget as an example.
+    data: Json
```

Comment:
I mean... who am I to complain.. this is literally what the spec says, but I wish this was more constrained. Like we know it will be a `SomethingBuildTarget` _right_? I can already see myself wanting tighter contraints than what the spec calls for, but I imagine that might be dangerous if we loosen up in the future and allow more things in `data`. But I wish `data` here could just be an `ADT` of like `BuildTargetData` where we had `ScalaBuildTarget`, etc as members.

Reply:
> But I wish data here could just be an ADT of like BuildTargetData where we had ScalaBuildTarget, etc as members. 

Right, so this is a very open question and there are several ways to approach it, but in the context of BSP, I believe no approach is perfect and will be subject to trade-offs, in particular in light of the following problems : 

#### Problem 1 : the BSP is an open protocol 

Doing what you ask means unavoidably creating a stronger bias towards Scala and Java, because suddenly you have stronger constraints for those things than you have for the rest. 

So when you say "Like we know it will be a SomethingBuildTarget right ?", I think the answer is "no, you don't know, because you've made BSP an open protocol as opposed to a closed one". 

It's basically like the difference between sealed traits and open-inheritance. You're asking the question implying that you want sealed-traits in a context where the decision was very much to leave the data unsealed so that the protocol could be re-used in other contexts than Scala/Java (unless I'm massively mis-understanding the scope of BSP) 

#### Problem 2 : the existing data types as they are defined in the BSP are ... not great 

The data field and discriminator are inlined with the rest of the information (like capabilities) and whatnot. This breaks the inductive nature of encoders/decoders : if we do what you want the way you want it, instead of code-generators being able to simply defer to macro-based derivation provided out of the box by Json libraries, the code-generators suddenly have to contain logic to ensure that the data is encoded correctly. 

So an alternative would be that the whole `BuildTarget` would become an ADT, where each of its member would abide by a shared interface and where only the `data` field would change. This is certainly possible to encode in Smithy, but it means the source of truth becomes less intuitive. 

#### Observation : 

There are a few knobs to play with : you can find ways of encoding things such that the spec would still be naive and give a pretty good idea of what the end json looks like, whilst tuning what the code-generators do to add more constraints when you use bsp4s (or bsp4j). 

To give you some intuition of what I mean : in Scala, you can easily decouple the data definition from the Json encoders/decoders. For instance, you can have `case class FooBar(foo: String, bar: String)` that could be encoded in json as `fooValue@barValue`. 

When people read `case class Foobar(...)`, what they expect the json will look like is 

```json 
{
  "foo" : "fooValue", 
  "bar" : "barValue"  
} 
```

Now say you need to retrofit some json value that looks like `fooValue@barValue` in an IDL. What do you do ? Do you say : 

1. "this is a string and it must look like this", or 
2. "it's a structure, but I'm gonna annotate it this special way to let everyone know it ought to be encoded this way" 

And that's the fundamental trade-off : if you go with 1, you make the spec precise but easier to implement. If you go with 2, you make the spec more precise but harder to implement. 
 
#### Conclusion : 

You need to be **extremely precise about what you're asking for**, and understand the trade-offs. I purposefully adhere to the spec in a somewhat oblivious manner because it's already gonna be a leap for the BSP maintainers to get used to the smithy language and metamodel and going with a fancy encoding for these things would make the leap even longer.

Reply:
Here are some more details that can fuel the discussion : 

* Smithy4s uses smithy-unions to represent sealed-coproducts, and provides 3 ways of encoding unions out of the box : https://disneystreaming.github.io/smithy4s/docs/codegen/unions
* AWS says that in the context of their protocols, unions are always encoded as tagged unions (read the link above to understand the difference). Arguably tagged unions are the best encoding for unions, in that this encoding can be parsed with no backtracking, and is the most flexible in terms in what you can put in the alternatives. So AWS is not wrong in how it tries to be super-constraining with unions.

Reply:
Alright, thanks the extra context here is super helpful.